### ISSUE TITLE
hist2d with log xy axis

### ISSUE THREAD

**Issue by jankoeh** [matplotlib/matplotlib#4615]
I often create 2d histograms with log x or y axis.
The `hist2d` function uses the `pcolorfast` method which can not handle log xy axes.

The following code explains this issue:

``` python
from pylab import *
axis = logspace(-1,0, 10)
hist2d(rand(100), rand(100), bins=(axis,axis)) #this still looks as it is supposed to look
xscale('log') #this scales only the ticks, but not the histogram
```

This could be fixed by using `pcolormesh` instead of `pcolorfast` in `axis.py`.
If there is no specific reason for using `pcolorfast` instead of `pcolormesh` I would propose to change this.

I'd be happy to create a fork and submit a pull request (it's only one line ...).

Thanks!

**Comment by tacaswell** [comment]
@jankoeh Is there a major performance difference when rendering the figures?  

A PR is the easiest way for us to understand and evaluate the change,

**Comment by efiring** [comment]
Pcolorfast uses AxesImage machinery for a uniform Cartesian grid; PcolorImage (derived from NonuniformImage) for an irregular rectangular grid; and pcolormesh for everything else.  I have no objection to switching hist2d to use pcolormesh; I very much doubt anyone will notice the reduced rendering speed, because hist2d is not going to be generating very dense grids.  Ideally, pcolorfast might use PcolorImage for log scales, but this would require considerable change to the code.  It would have to become an Artist so that it could decide in a draw() method which of the 3 drawing methods to use.

**Comment by jankoeh** [comment]
I created a pull request - I was unsure whether to use the master or maintenance branch, I used master.

Concerning the performance difference, I did not do any in depths analysis, but I'm frequently using `pcolormesh` to plot `numpy.histogram2d` data with several 100 x several 100 bins. So far I did not experience any performance issues.

**Comment by tacaswell** [comment]
This is also related to #4152

**Comment by dstansby** [comment]
Closed by #9987
